src/main/host/descriptor/pipe.rs

```diff
@@ -69,14 +69,19 @@ impl PipeFile {
             return SyscallReturn::Error(nix::errno::Errno::EBADF);
         }
 
+        let bytes = match bytes {
```

Comment:
Optional throughout: more succinct as

```
let bytes = bytes.ok_or(SyscallReturn::Error(nix::errno::Errno::EFAULT))
```

Or maybe avoid constructing the error in the ok case (but probably a wash in this case)
```
let bytes = bytes.ok_or_else(|| SyscallReturn::Error(nix::errno::Errno::EFAULT))
```

Reply:
This line replaces an `Option<&[u8]>` with a `Result<&[u8], SyscallReturn>`, but we still need to unpack the type to either return the `SyscallReturn`, or destructure the enum to get the `&[u8]`.

Reply:
Ah right. Maybe we ought to break the Error state out of SyscallReturn so that we can use Result types throughout, and here use the `?` operator (We'd need to define `From<nix::errno::Errno>` for our error type, assuming we didn't use that directly as our error type)

Seems orthogonal to this PR though, so nevermind for now :)

Reply:
My reasoning for not doing it that way is that `SyscallReturn::Error` represents an error state for the plugin and not Shadow. So having a Shadow function return a `Result::Err` when it's not actually a Shadow error might be confusing. And then, how do we return an actual Shadow error if we're already using a `Result::Err` to return the plugin error state?

I'm not against doing it this way, especially if it cuts down on the amount of code. It's just not clear to me which way is better in the long term.

Reply:
> My reasoning for not doing it that way is that SyscallReturn::Error represents an error state for the plugin and not Shadow.

I think that's ok. Right now there is no mechanism for returning a Shadow error, so we wouldn't actually lose anything through this transformation.

> So having a Shadow function return a Result::Err when it's not actually a Shadow error might be confusing.

Somewhat subjective, but IMO using `Result:Err` for this sort of use-case is reasonably idiomatic.

> And then, how do we return an actual Shadow error if we're already using a Result::Err to return the plugin error state?

We'd only need to do that in the case of some kind of recoverable Shadow error. So far though most Shadow errors aren't recoverable, and we just want to `panic` (well, `error`) right where the error happened rather than bubbling it up through the call stack.

If we really wanted to be able to represent Shadow errors though, we could use an error type like:

```
enum SyscallError {
  PluginError(nix::errno::Errno),
  ShadowError1(...),
  ShadowError2(...),
}
```